### PR TITLE
[WIP] Fixes for 0.7/1.0

### PR DESCRIPTION
**Pull request by rejuvyesh**

### CODE REVIEW

**Comment by Shushman** [comment]
Update
- All syntax errors are removed except for two hard-to-trace ones:
```
┌ Warning: Duplicate variable names are deprecated: pass makeunique=true to add a suffix automatically.
│   caller = ip:0x0
└ @ Core :-1
┌ Warning: the behavior of `A[I...] .= X` with scalar indices will change in the future. Use `A[I...] = X` instead.
│   caller = ip:0x0
└ @ Core :-1
```

- All tests pass except two - `test_inference.jl` has an actual testing error
```
Test Failed at /Users/shushmanchoudhury/.julia/dev/BayesNets/test/test_inference.jl:62
  Expression: isapprox((ϕ[:c => 1]).potential[1], 1.0, atol=0.02)
   Evaluated: isapprox(0.42500000000000004, 1.0; atol=0.02)
```

- `test_docs.jl` fails due to `NBInclude.jl` calling `srand`

**Comment by tawheeler** [comment]
Will do

**Comment by tawheeler** [comment]
So I installed julia 1.0.0, ran `] develop BayesNets`, navigated to the repo, checked out `jkg/fixes-0.7`, ran `] test BayesNets`, and am not getting an error that SpecialFunctions fails to compile correctly.

On julia 0.7, testing the branch immediately fails with `ERROR: MethodError: no method matching haskey(::Array{String,1}, ::String)`.

**Comment by Shushman** [comment]
Okay so a couple of things:
- For the `test BayesNets` on 1.0 failing, I can't reproduce that. When I test it, I get `ERROR: LoadError: LoadError: LoadError: UndefVarError: reducedim not defined`

- For julia 0.7, your error is probably related to this issue with `TikzGraphs` (https://github.com/JuliaTeX/TikzGraphs.jl/issues/25)

- For `TikzPictures`, it did have issues with `mimewritable` (which should now be `showable`), but I believe I addressed that in https://github.com/JuliaTeX/TikzPictures.jl/commit/7bf4fe963d9805867af84945f6c8cbc0dfb84ec1

**Comment by Shushman** [comment]
Oh I got it! If you just did `dev BayesNets`, it still pulls the _registered_ versions of the dependencies (like `TikzGraphs`) which do not work with 1.0 yet, since they are currently in progress. You need to `dev` the various dependencies and check out their branches with 0.7 fixes (sometimes that might be `master`), before you can test `BayesNets` directly. For `BayesNets` I think it is only `TikzGraphs` that needs to be pulled, but for `TikzGraphs` you need to `dev TikzPictures` first

**Comment by rejuvyesh** [comment]
```
Test Summary:    | Pass  Fail  Error  Total
bn               |  472     2      1    475
  utils          |    6                   6
  cpds           |   78                  78
  tables         |   13                  13
  factors        |   41                  41
  bayesnets      |   41                  41
  gibbs          |   45                  45
  sampling       |    8                   8
  inference      |   99     2           101
  learning       |   30                  30
  io             |   28                  28
  ndgrid         |    8                   8
  discrete bayes |   43                  43
  gen bn         |   32                  32
  docs           |                 1      1
ERROR: LoadError: Some tests did not pass: 472 passed, 2 failed, 1 errored, 0 broken.
```

File: REQUIRE [hunk 1]
```diff
@@ -8,3 +8,4 @@ LightXML 0.3.0
 Reexport 0.0.3
 Parameters 0.5.0
 DataFrames 0.11.2
+StatsBase
```
Comment (tawheeler) [inline REQUIRE]: Bump to julia 0.7?

**Comment by mykelk** [comment]
I think the `.travis.yml` needs to be updated to test 0.7 and 1.0.

**Comment by tawheeler** [comment]
Finally able to reproduce test results. Was on julia0.7.0-beta2 and not the released 0.7.

**Comment by tawheeler** [comment]
Not sure how to configure travis to use the project files.

**Comment by mykelk** [comment]
It looks like the way to do it now is something like how they do it [here](https://github.com/JuliaDocs/Documenter.jl).

**Comment by rejuvyesh** [comment]
Current travis error makes no sense.

File: src/learning.jl [hunk 2]
```diff
@@ -2,35 +2,35 @@ function Distributions.fit(::Type{BayesNet}, data::DataFrame, dag::DAG, cpd_type
...
 
 function _get_dag(data::DataFrame, edges::Tuple{Vararg{Pair{NodeName, NodeName}}})
     varnames = names(data)
     dag = DAG(length(varnames))
     for (a,b) in edges
-        i = findfirst(varnames, a)
-        j = findfirst(varnames, b)
+        i = something(findfirst(isequal(a), varnames), 0)
+        j = something(findfirst(isequal(b), varnames), 0)
         add_edge!(dag, i, j)
```
Comment (tawheeler) [inline src/learning.jl]: This is going to fail even if i and j are set to 0. Might as well drop the `something` calls.

**Comment by rejuvyesh** [comment]
TODO things for 1.0:

- No `reducedims` anymore. Do we want to follow new Julia standard of using `reduce(...; dims=<>)` instead?
- In `factors_access.jl`, `Base.setindex` runs into issues because the value is sometimes integer, other times array.

**Comment by rejuvyesh** [comment]
@mykelk and @tawheeler: All tests pass. But both the earlier fixes are ugly and can probably be fixed by refactoring other things. Let me know if we should look into it or current fixes are good enough.

**Comment by tawheeler** [comment]
> Let me know if we should look into it or current fixes are good enough.

I'd release a package that works and then patch it if we find it should/can be improved.

**Pull request merged**